compressors/lz77.py

```diff
@@ -0,0 +1,705 @@
+"""
+LZ77 works by finding matches in the past and encoding the match lengths and offsets.
+Symbols not part of matches are directly stored as "literals". These streams are 
+then entropy coded. LZ77 forms the basis of popular compressors like gzip and zstd.
+Theoretically LZ77 is a universal compressor, that is, it asymptotically achieves the
+entropy rate of any stationary process.
+
+This is a simplified LZ77 encoder deriving ideas from: 
+- LZ77 modern implementation: https://glinscott.github.io/lz/index.html 
+- DEFLATE: https://www.rfc-editor.org/rfc/rfc1951 
+- ZSTD: https://github.com/facebook/zstd/blob/dev/doc/zstd_compression_format.md
+
+The encoder has 2 parameters: 
+- min_match_length: this determines the minimum
+allowed match length. The same length is also used for indexing into the dict.
+Very low values lead to slow execution. In terms of compression, low values can
+lead to suboptimal matches but high values can lead to missed matches. So there
```

Comment:
why should low values lead to sub-optimal matches? (because there are too many?)

Reply:
(I see, because of the next paramter)

Reply:
yes